crates/storage/src/lazy/mod.rs

```diff
@@ -18,4 +18,258 @@
+where
+    KeyType: StorageKey,
+{
+    fn fmt(&self, f: &mut core::fmt::Formatter) -> core::fmt::Result {
+        f.debug_struct("Lazy").field("key", &KeyType::KEY).finish()
+    }
+}
+
+impl<V, KeyType> Lazy<V, KeyType>
+where
+    V: Storable,
+    KeyType: StorageKey,
+{
+    /// Reads the `value` from the contract storage.
+    pub fn get(&self) -> Option<V> {
+        match ink_env::get_contract_storage::<Key, V>(&KeyType::KEY) {
+            Ok(Some(value)) => Some(value),
+            _ => None,
+        }
+    }
+}
+
+impl<V, KeyType> Lazy<V, KeyType>
+where
+    V: Storable + Default,
+    KeyType: StorageKey,
+{
+    /// Reads the `value` from the contract storage.
+    ///
+    /// Returns `Default::default()` if no `value` exists.
+    pub fn get_or_default(&self) -> V {
+        match ink_env::get_contract_storage::<Key, V>(&KeyType::KEY) {
+            Ok(Some(value)) => value,
+            _ => Default::default(),
+        }
+    }
+}
+
+impl<V, KeyType> Lazy<V, KeyType>
+where
+    V: Storable,
+    KeyType: StorageKey,
+{
+    /// Writes the given `value` to the contract storage.
+    pub fn set(&mut self, value: &V) {
+        push_storage(&KeyType::KEY, value);
+    }
+}
+
+impl<V, KeyType> Storable for Lazy<V, KeyType>
+where
+    KeyType: StorageKey,
+{
+    #[inline(always)]
+    fn encode<T: Output + ?Sized>(&self, _dest: &mut T) {}
+
+    #[inline(always)]
+    fn decode<I: Input>(_input: &mut I) -> Result<Self, Error> {
+        Ok(Default::default())
+    }
+}
+
+impl<V, Key, InnerKey> Item<Key> for Lazy<V, InnerKey>
+where
+    Key: StorageKey,
+    InnerKey: StorageKey,
+    V: Item<Key>,
+{
+    type Type = Lazy<V::Type, Key>;
+    type PreferredKey = InnerKey;
+}
+
+impl<V, KeyType> StorageKey for Lazy<V, KeyType>
+where
+    KeyType: StorageKey,
+{
+    const KEY: Key = KeyType::KEY;
+}
+
+#[cfg(feature = "std")]
+const _: () = {
+    use crate::traits::StorageLayout;
+    use ink_metadata::layout::{
+        Layout,
+        LayoutKey,
+        RootLayout,
+    };
+
+    impl<V, KeyType> StorageLayout for Lazy<V, KeyType>
+    where
+        V: StorageLayout + scale_info::TypeInfo + 'static,
+        KeyType: StorageKey + scale_info::TypeInfo + 'static,
+    {
+        fn layout(_: &Key) -> Layout {
+            Layout::Root(RootLayout::new(
+                LayoutKey::from(&KeyType::KEY),
+                <V as StorageLayout>::layout(&KeyType::KEY),
+            ))
+        }
+    }
+};
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::traits::ManualKey;
+
+    #[test]
+    fn set_and_get_work() {
+        ink_env::test::run_test::<ink_env::DefaultEnvironment, _>(|_| {
+            let mut storage: Lazy<u8, ManualKey<123>> = Lazy::new();
+            storage.set(&2);
+            assert_eq!(storage.get(), Some(2));
+
+            Ok(())
+        })
+        .unwrap()
+    }
+
+    #[test]
+    fn set_and_get_work_for_two_lazy_with_same_manual_key() {
```

Comment:
We should make a note somewhere, maybe as part of the doc comments of `Lazy` as to why this is powerful but also dangerous.

I can picture cases where developers copy paste some fields in a struct and end up with two fields that modify the same area in storage inadvertently